packages/devtools_app/lib/src/charts/chart.dart

```diff
@@ -0,0 +1,588 @@
+        labelIndex < chartController.getLabelsCount();
+        labelIndex++) {
+      final timestamp = chartController.getLabelTimestampByIndex(labelIndex);
+      if (timestamp != null) {
+        final xCoord = chartController.timestampXCanvasCoord(timestamp);
+        drawXTick(canvas, timestamp, xCoord, axis, displayTime: true);
+      }
+    }
+
+    // Rescale Y-axis to max visible Y range.
+    chartController.resetYMaxValue(visibleYMax);
+
+    // Draw Y-axis ticks and labels.
+    if (chartController.displayYLabels) {
+      drawYTicks(canvas, chartController, axis);
+    }
+
+    drawTitle(canvas, size, chartController.title);
+  }
+
+  void clipChart(Canvas canvas, {ClipOp op = ClipOp.intersect}) {
+    final leftSideSide = chartController.xCanvasChart;
+    final topChartSide = chartController.yCanvasChart;
+    final r = Rect.fromLTRB(
+      leftSideSide,
+      topChartSide,
+      chartController.canvasChartWidth +
+          leftSideSide -
+          chartController.xPaddingRight,
+      topChartSide + chartController.canvasChartHeight,
+    );
+
+    canvas.clipRect(r, clipOp: op);
+  }
+
+  // TODO(terry): Use drawText?
+  void drawTitle(Canvas canvas, Size size, String title) {
+    final tp = createText(title, 1.5);
+    tp.paint(canvas, Offset(size.width / 2 - tp.width / 2, 0));
+  }
+
+  void drawAxes(Canvas canvas, Size size, Paint axis) {
+    // Top line of chart.
+    canvas.drawLine(
+      Offset(
+        chartController.xCanvasChart,
+        chartController.yCanvasChart,
+      ),
+      Offset(
+        chartController.canvasChartWidth +
+            chartController.xCanvasChart -
+            chartController.xPaddingRight,
+        chartController.yCanvasChart,
+      ),
+      axis,
+    );
+
+    // Left-side of chart
+    canvas.drawLine(
+      Offset(
+        chartController.xCanvasChart,
+        chartController.yCanvasChart,
+      ),
+      Offset(
+        chartController.xCanvasChart,
+        chartController.yCanvasChart + chartController.canvasChartHeight,
+      ),
+      axis,
+    );
+
+    // Bottom line of chart.
+    canvas.drawLine(
+      Offset(
+        chartController.xCanvasChart,
+        chartController.yCanvasChart + chartController.canvasChartHeight,
+      ),
+      Offset(
+        chartController.canvasChartWidth +
+            chartController.xCanvasChart -
+            chartController.xPaddingRight,
+        chartController.yCanvasChart + chartController.canvasChartHeight,
+      ),
+      axis,
+    );
+  }
+
+  /// Separated out from drawAxis because we don't know range until plotted.
+  // TODO(terry): Consider combining into drawAxis function.
+  void drawYTicks(Canvas canvas, ChartController chartController, Paint axis) {
+    final xCoord = chartController.xCanvasChart;
+
+    final yScale = chartController.yScale;
+
+    for (var labelIndex = yScale.labelTicks; labelIndex >= 0; labelIndex--) {
+      final unit = pow(10, yScale?.labelUnitExponent).toInt();
```

Comment:
maybe make this .floor() to make it clear that you intentionally want to round down?

Reply:
Done.